File: backend/services/block_rules.py
"""
Luật tương thích khối — "luật sân chơi" (nguồn sự thật phía backend).

⚠️ GIỮ ĐỒNG BỘ với frontend/src/config/blockRules.js.

Mở rộng: khối mới chỉ cần gắn tag vào BLOCK_CAPS; luật viết theo TAG.
"""
import json

# Tag năng lực của từng loại khối
BLOCK_CAPS = {
    "input_vars": {"interactive": True},       # chờ người nhập giữa chừng
    "telegram_listener": {"listener": True},   # chạy nền chờ tin nhắn
}

# Bảng luật (xem giải thích ở file FE)
WF_RULES = [
    {"when": "interactive", "forbid_tag": "listener",
     "msg": 'Khối "Biến đầu vào" (chờ người nhập) không dùng chung với Telegram Listener trong cùng một workflow.'},
    {"when": "interactive", "disables_feature": "scheduler",
     "msg": 'Workflow có khối "Biến đầu vào" (chờ người nhập) nên không thể đặt lịch chạy tự động.'},
]


def _nodes_from_graph(graph_json):
    if not graph_json:
        return []
    try:
        g = json.loads(graph_json) if isinstance(graph_json, str) else graph_json
        return g.get("nodes", []) or []
    except Exception:
        return []


def _collect_tags(nodes):
    tags = set()
    for n in nodes or []:
        btype = (n.get("data") or {}).get("type")
        caps = BLOCK_CAPS.get(btype)
        if caps:
            for k, v in caps.items():
                if v:
                    tags.add(k)
    return tags


def validate_workflow(graph_json):
    """Trả {"ok": bool, "violations": [msg...], "disabled_features": set()}."""
    tags = _collect_tags(_nodes_from_graph(graph_json))
    violations, disabled = [], set()
    for r in WF_RULES:
        if r["when"] not in tags:
            continue
        if r.get("forbid_tag") and r["forbid_tag"] in tags:
            violations.append(r["msg"])
        if r.get("disables_feature"):
            disabled.add(r["disables_feature"])
    return {"ok": len(violations) == 0, "violations": violations, "disabled_features": disabled}


def is_feature_disabled(graph_json, feature):
    return feature in validate_workflow(graph_json)["disabled_features"]

File: frontend/src/config/blockRules.js
/**
 * Luật tương thích khối — "luật sân chơi".
 *
 * ⚠️ GIỮ ĐỒNG BỘ với backend/services/block_rules.py (BE là nguồn sự thật; file
 * này chỉ để chặn thao tác + hiển thị lý do cho người dùng ở FE).
 *
 * Cách mở rộng: thêm khối mới chỉ cần gắn tag năng lực vào BLOCK_CAPS; luật viết
 * theo TAG nên không phải sửa từng cặp khối.
 */

// Tag năng lực của từng loại khối
export const BLOCK_CAPS = {
  input_vars: { interactive: true },      // chờ người nhập giữa chừng
  telegram_listener: { listener: true },  // chạy nền chờ tin nhắn
}

// Bảng luật ràng buộc theo tag
//  - when: tag điều kiện có mặt trong workflow
//  - forbidTag: nếu đồng thời có tag này → xung khắc (không cho cùng workflow)
//  - disablesFeature: tính năng bị tắt cho workflow đó (vd 'scheduler')
export const WF_RULES = [
  {
    when: 'interactive', forbidTag: 'listener',
    msg: 'Khối "Biến đầu vào" (chờ người nhập) không dùng chung với Telegram Listener trong cùng một workflow.',
  },
  {
    when: 'interactive', disablesFeature: 'scheduler',
    msg: 'Workflow có khối "Biến đầu vào" (chờ người nhập) nên không thể đặt lịch chạy tự động.',
  },
]

// Tập tag đang hiện diện trong danh sách node
function collectTags(nodes) {
  const tags = new Set()
  for (const n of nodes || []) {
    const caps = BLOCK_CAPS[n?.data?.type]
    if (caps) for (const k of Object.keys(caps)) if (caps[k]) tags.add(k)
  }
  return tags
}

// Kiểm 1 workflow → { ok, violations:[{msg}], disabledFeatures:Set<string> }
export function validateWorkflow(nodes) {
  const tags = collectTags(nodes)
  const violations = []
  const disabledFeatures = new Set()
  for (const rule of WF_RULES) {
    if (!tags.has(rule.when)) continue
    if (rule.forbidTag && tags.has(rule.forbidTag)) violations.push({ msg: rule.msg })
    if (rule.disablesFeature) disabledFeatures.add(rule.disablesFeature)
  }
  return { ok: violations.length === 0, violations, disabledFeatures }
}

// Có được phép THÊM khối newType vào graph hiện tại không → { ok, msg }
export function canAddBlock(nodes, newType) {
  const existing = collectTags(nodes)
  const newCaps = BLOCK_CAPS[newType] || {}
  const newTags = Object.keys(newCaps).filter(k => newCaps[k])
  for (const rule of WF_RULES) {
    if (!rule.forbidTag) continue
    const addWhen = newTags.includes(rule.when) && existing.has(rule.forbidTag)
    const addForbid = newTags.includes(rule.forbidTag) && existing.has(rule.when)
    if (addWhen || addForbid) return { ok: false, msg: rule.msg }
  }
  return { ok: true }
}

export function isFeatureDisabled(nodes, feature) {
  return validateWorkflow(nodes).disabledFeatures.has(feature)
}
